validate.sh

```diff
@@ -0,0 +1,183 @@
+#!/bin/bash
+
+RUNS_PER_THREAD=100000
+MAX_PTHREADS=$(nproc)
+BIN_PATH="./build/bin"
+RES_FILE="res.txt"
+BUILD_LOG="build.log"
+STEP_SELETCTED=0
+STEP_THREADS=1
+FIBERS=""
+NUMA_NODES=""
+PASS_STATUS=1
+
+function usage()
+{
+    echo -e "Usage: ./validate.sh OPTION=NUM ...";
+    echo -e "This script compiles the sources in DEBUG mode and validates the correctnes of some of the provided concurrent objects.";
+    echo -e ""
+    echo -e "The following options are available."
+    echo -e "-t, --max_threads \t set the maximum number number of POSIX threads to be used in the last set of iterations of the benchmark, default is the number of system_cpus"
+    echo -e "-s, --step \t set the step (extra number of POSIX threads to be used) in succesive set of iterations of the benchmark, default is number of system_cpu/8 or 1"
+    echo -e "-f, --fibers  \t set the number of fibers (user-level threads) per posix thread."
+    echo -e "-r, --runs    \t set the total number of operations executed by each thread of each benchmark, default is ${RUNS_PER_THREAD}"
+    echo -e "-w, --max_work\t set the amount of workload (i.e. dummy loop iterations among two consecutive operations of the benchmarked object), default is 64"
+    echo -e ""
+    echo -e "-h, --help    \t displays this help and exits"
+    echo -e ""
+}
+
+COLOR_PASS="[ \e[32mPASS\e[39m ]"
+COLOR_FAIL="[ \e[31mFAIL\e[39m ]"
+
+declare -a uobjects=("ccsynchbench.run"                     "dsmsynchbench.run" "hsynchbench.run" "oscibench.run"      "simbench.run")
+declare -a queues=(  "ccqueuebench.run" "clhqueuebench.run" "dsmqueuebench.run" "hqueuebench.run" "osciqueuebench.run" "simqueuebench.run")
+declare -a stacks=(  "ccstackbench.run" "clhstackbench.run" "dsmstackbench.run" "hstackbench.run" "oscistackbench.run" "simstackbench.run")
+
+declare -a validate_objects=()
+
+validate_objects+=${uobjects[@]}
+validate_objects+=${queues[@]}
+validate_objects+=${stacks[@]}
+
+if [ "$1" = "--help" ] || [ "$1" = "-h" ]; then
+    usage;
+    exit;
+fi
+
+while [ "$1" != "" ]; do
+    PARAM=`echo $1 | awk -F= '{print $1}'`
+    VALUE=`echo $1 | awk -F= '{print $2}'`
+    case $PARAM in
+        -h | --help)
+            usage;
+            exit;
+            ;;
+        -s | --step_threads)
+            STEP_THREADS=$VALUE
+            STEP_SELETCTED=1
+            ;;
+        -t | --max_threads)
+            MAX_PTHREADS=$VALUE
+            ;;
+        -f | --fibers)
+            FIBERS="-f $VALUE"
+            ;;
+        -w | --max_work)
+            WORKLOAD="-w $VALUE"
```

Comment:
where is the default value of `64` specified?

Reply:
This is defined in config.h.

Reply:
Well this kind of dependencies is hard to track. If config.h changes then this file will be out of sync...
I wonder how can we alleviate this.

Reply:
One solution is to remove this option from the validation it say something like:
"Default value is specified in config.h"

Reply:
OK, let's follow a bit different approach.
I'm going to explicitly set this to 64 in the beginning of the bench.sh.
Thus, the 64 should be valid for the scripts, but may be different for the executables.
Are you agree wit this?

Reply:
I am fine with this approach since bench.sh is a test and can have a different default value for specific options.